### ISSUE TITLE
Unable to disable verification on vbmeta_a

### ISSUE THREAD

**Issue by 4aiman** [bkerler/mtkclient#1240]
Android 13, MT6789 + windows 11 + mtkclient pulled on 10/05/24

Trying to root the device but stuck at step 8:
![](https://github.com/user-attachments/assets/cac7dfeb-d247-4b32-a00c-8f87f3473f42)

```
...........
Port - Device detected :)
Preloader -     CPU:                    MT6789(MTK Helio G99)
Preloader -     HW version:             0x0
Preloader -     WDT:                    0x10007000
Preloader -     Uart:                   0x11002000
Preloader -     Brom payload addr:      0x100a00
Preloader -     DA payload addr:        0x201000
Preloader -     Var1:                   0xa
Preloader - Disabling Watchdog...
Preloader - HW code:                    0x1208
Preloader - Target config:              0xe0
Preloader -     SBC enabled:            False
Preloader -     SLA enabled:            False
Preloader -     DAA enabled:            False
Preloader -     SWJTAG enabled:         False
Preloader -     EPP_PARAM at 0x600 after EMMC_BOOT/SDMMC_BOOT:  False
Preloader -     Root cert required:     False
Preloader -     Mem read auth:          True
Preloader -     Mem write auth:         True
Preloader -     Cmd 0xC8 blocked:       True
Preloader - Get Target info
Preloader - BROM mode detected.
Preloader -     HW subcode:             0x8a00
Preloader -     HW Ver:                 0xca00
Preloader -     SW Ver:                 0x0
Preloader - ME_ID:                      556B56CC473F6191882543187FDBE1FF
Preloader - SOC_ID:                     2DAE8FBBDB78BFB8BDA730338E236311B0BE5EF980A193D5BAD51CAD335B1EDA
DaHandler - Device is unprotected.
DaHandler - Device is in BROM-Mode. Bypassing security.
DaHandler
DaHandler - [LIB]: Failed to dump preloader from ram, provide a valid one via --preloader option
DAXML - Uploading xflash stage 1 from MTK_DA_V6.bin
XmlFlashExt - Patching da2 ...
XmlFlashExt - Patched read_register / write_register
XmlFlashExt - Patched write partitions / allow_read / allow_write
XmlFlashExt - Patched hash binding
DAXML - Successfully uploaded stage 1, jumping ..
Preloader - Jumping to 0x200000
Preloader - Jumping to 0x200000: ok.
DAXML - Stage 1 successfully loaded.
DAXML - Uploading stage 2...
Progress: |██████████| 100.0% Written (0x286/0x286, ) 3.05 MB/s
DAXML - Successfully uploaded stage 2.
DAXML - Successfully uploaded stage 2
DAXML - SLA is disabled
DAXML - DA XML Extensions successfully loaded.
DaHandler - Dumping partition "vbmeta_a"
Progress: |██████████| 100.0% Read (0x4000/0x4000, ) 5.42 MB/s21 MB/s
DaHandler - Patching vbmeta"
DaHandler - Patching verification + verity
DaHandler - Writing partition "vbmeta_a"
Traceback (most recent call last):
  File "F:\Mobile\mtkclient\mtk.py", line 1021, in <module>
    main()
  File "F:\Mobile\mtkclient\mtk.py", line 1017, in main
    mtk = Main(args).run(parser)
  File "F:\Mobile\mtkclient\mtkclient\Library\mtk_main.py", line 684, in run
    da_handler.handle_da_cmds(mtk, cmd, self.args)
  File "F:\Mobile\mtkclient\mtkclient\Library\DA\mtk_da_handler.py", line 900, in handle_da_cmds
    self.da_vbmeta(vbmode=vbmode)
  File "F:\Mobile\mtkclient\mtkclient\Library\DA\mtk_da_handler.py", line 202, in da_vbmeta
    if self.mtk.daloader.writeflash(addr=rpartition.sector * self.config.pagesize,
  File "F:\Mobile\mtkclient\mtkclient\Library\DA\mtk_daloader.py", line 305, in writeflash
    return self.da.writeflash(addr=addr, length=length, filename=filename, offset=offset,
  File "F:\Mobile\mtkclient\mtkclient\Library\DA\xml\xml_lib.py", line 970, in writeflash
    data = fh.read(length)
AttributeError: 'NoneType' object has no attribute 'read'
```

**Comment by 4aiman** [comment]
I've patched `mtk_da_handler.py` to see what's in there in the patched vbmeta:
![](https://github.com/user-attachments/assets/f90bb12a-6432-4fe5-b044-72a3a0fd091b)
but the resulting file is just a blank dummy file of the same size as the original vb_meta. 
Is that right?
If it is, then why is it "unflashable"?

**Comment by 4aiman** [comment]
I've also tried to flash that empty file on my own, but it failed again:
```
PS F:\Mobile\mtkclient> ./mtk.bat w boot_a,vbmeta_a magisk_patched-27000_boota_BoEq9.img,patched_vbmeta.bin
MTK Flash/Exploit Client Public V2.0.1 (c) B.Kerler 2018-2024

Preloader - Status: Waiting for PreLoader VCOM, please reconnect mobile to brom mode

Port - Hint:

Power off the phone before connecting.
For brom mode, press and hold vol up, vol dwn, or all hw buttons and connect usb.
For preloader mode, don't press any hw button and connect usb.
If it is already connected and on, hold power for 10 seconds to reset.


...........

Port - Hint:

Power off the phone before connecting.
For brom mode, press and hold vol up, vol dwn, or all hw buttons and connect usb.
For preloader mode, don't press any hw button and connect usb.
If it is already connected and on, hold power for 10 seconds to reset.


.........Port - Device detected :)
Preloader -     CPU:                    MT6789(MTK Helio G99)
Preloader -     HW version:             0x0
Preloader -     WDT:                    0x10007000
Preloader -     Uart:                   0x11002000
Preloader -     Brom payload addr:      0x100a00
Preloader -     DA payload addr:        0x201000
Preloader -     Var1:                   0xa
Preloader - Disabling Watchdog...
Preloader - HW code:                    0x1208
Preloader - Target config:              0xe0
Preloader -     SBC enabled:            False
Preloader -     SLA enabled:            False
Preloader -     DAA enabled:            False
Preloader -     SWJTAG enabled:         False
Preloader -     EPP_PARAM at 0x600 after EMMC_BOOT/SDMMC_BOOT:  False
Preloader -     Root cert required:     False
Preloader -     Mem read auth:          True
Preloader -     Mem write auth:         True
Preloader -     Cmd 0xC8 blocked:       True
Preloader - Get Target info
Preloader - BROM mode detected.
Preloader -     HW subcode:             0x8a00
Preloader -     HW Ver:                 0xca00
Preloader -     SW Ver:                 0x0
Preloader - ME_ID:                      556B56CC473F6191882543187FDBE1FF
Preloader - SOC_ID:                     2DAE8FBBDB78BFB8BDA730338E236311B0BE5EF980A193D5BAD51CAD335B1EDA
DaHandler - Device is unprotected.
DaHandler - Device is in BROM-Mode. Bypassing security.
DaHandler
DaHandler - [LIB]: Failed to dump preloader from ram, provide a valid one via --preloader option
DAXML - Uploading xflash stage 1 from MTK_DA_V6.bin
XmlFlashExt - Patching da2 ...
XmlFlashExt - Patched read_register / write_register
XmlFlashExt - Patched write partitions / allow_read / allow_write
XmlFlashExt - Patched hash binding
DAXML - Successfully uploaded stage 1, jumping ..
Preloader - Jumping to 0x200000
Preloader - Jumping to 0x200000: ok.
DAXML - Stage 1 successfully loaded.
DAXML - Uploading stage 2...
Progress: |██████████| 100.0% Written (0x286/0x286, ) 3.05 MB/s
DAXML - Successfully uploaded stage 2.
DAXML - Successfully uploaded stage 2
DAXML - SLA is disabled
DAXML - DA XML Extensions successfully loaded.
Progress: |██████████| 100.0% Written (0x20000/0x20000, ) 37.41 MB/s7 MB/s
Wrote magisk_patched-27000_boota_BoEq9.img to sector 174592 with sector count 16384.
DAXML
DAXML - [LIB]: Read flash isn't supported
DaHandler
DaHandler - [LIB]: Error: Couldn't detect partition: vbmeta_a
Available partitions:
```
And there was nothing after that last line.

Undeterred, I've tried yet again, with just the blank vbmeta_a:
```
PS F:\Mobile\mtkclient> ./mtk.bat w vbmeta_a patched_vbmeta.bin
MTK Flash/Exploit Client Public V2.0.1 (c) B.Kerler 2018-2024

Preloader - Status: Waiting for PreLoader VCOM, please reconnect mobile to brom mode

Port - Hint:

Power off the phone before connecting.
For brom mode, press and hold vol up, vol dwn, or all hw buttons and connect usb.
For preloader mode, don't press any hw button and connect usb.
If it is already connected and on, hold power for 10 seconds to reset.


...........

Port - Hint:

Power off the phone before connecting.
For brom mode, press and hold vol up, vol dwn, or all hw buttons and connect usb.
For preloader mode, don't press any hw button and connect usb.
If it is already connected and on, hold power for 10 seconds to reset.


..........Port - Device detected :)
Preloader -     CPU:                    MT6789(MTK Helio G99)
Preloader -     HW version:             0x0
Preloader -     WDT:                    0x10007000
Preloader -     Uart:                   0x11002000
Preloader -     Brom payload addr:      0x100a00
Preloader -     DA payload addr:        0x201000
Preloader -     Var1:                   0xa
Preloader - Disabling Watchdog...
Preloader - HW code:                    0x1208
Preloader - Target config:              0xe0
Preloader -     SBC enabled:            False
Preloader -     SLA enabled:            False
Preloader -     DAA enabled:            False
Preloader -     SWJTAG enabled:         False
Preloader -     EPP_PARAM at 0x600 after EMMC_BOOT/SDMMC_BOOT:  False
Preloader -     Root cert required:     False
Preloader -     Mem read auth:          True
Preloader -     Mem write auth:         True
Preloader -     Cmd 0xC8 blocked:       True
Preloader - Get Target info
Preloader - BROM mode detected.
Preloader -     HW subcode:             0x8a00
Preloader -     HW Ver:                 0xca00
Preloader -     SW Ver:                 0x0
Preloader - ME_ID:                      556B56CC473F6191882543187FDBE1FF
Preloader - SOC_ID:                     2DAE8FBBDB78BFB8BDA730338E236311B0BE5EF980A193D5BAD51CAD335B1EDA
DaHandler - Device is unprotected.
DaHandler - Device is in BROM-Mode. Bypassing security.
DaHandler
DaHandler - [LIB]: Failed to dump preloader from ram, provide a valid one via --preloader option
DAXML - Uploading xflash stage 1 from MTK_DA_V6.bin
XmlFlashExt - Patching da2 ...
XmlFlashExt - Patched read_register / write_register
XmlFlashExt - Patched write partitions / allow_read / allow_write
XmlFlashExt - Patched hash binding
DAXML - Successfully uploaded stage 1, jumping ..
Preloader - Jumping to 0x200000
Preloader - Jumping to 0x200000: ok.
DAXML - Stage 1 successfully loaded.
DAXML - Uploading stage 2...
Progress: |██████████| 100.0% Written (0x286/0x286, ) 0.00 MB/s
DAXML - Successfully uploaded stage 2.
DAXML - Successfully uploaded stage 2
DAXML - SLA is disabled
DAXML - DA XML Extensions successfully loaded.
Progress: |██████████| 100.0% Written (0x4000/0x4000, ) 38.03 MB/s
Wrote patched_vbmeta.bin to sector 57864 with sector count 2048.
```
I *think* it's "freezing" just like in #1239, but now in the command line interface.


However, even after all of that, Magisk app is showing there's no root:
![](https://github.com/user-attachments/assets/1ad9cc96-bf67-4a60-bc14-619e0bbbe215)

My device *does* have init_boot_a and init_boot_b, but Magicsk can't patch them.

**Comment by 4aiman** [comment]
Then I thought my device might be on the boot_b partition.
I re-did everything (patching vbmeta, then patching boot_b and flashing it all in sequence) and got myself a red state with a bootloop.

Seems something's wrong in this process, as the bootloader had been unlocked prior to patching

**Comment by bkerler** [comment]
There was a bug in the xml writeflash function, fixed it in [127c353](https://github.com/bkerler/mtkclient/commit/127c353b397bd2681f63d3b16fd5b3dfa369e8e6)

**Comment by 4aiman** [comment]
Thanks! It patches!

```
......Port - Device detected :)
Preloader -     CPU:                    MT6789(MTK Helio G99)
Preloader -     HW version:             0x0
Preloader -     WDT:                    0x10007000
Preloader -     Uart:                   0x11002000
Preloader -     Brom payload addr:      0x100a00
Preloader -     DA payload addr:        0x201000
Preloader -     Var1:                   0xa
Preloader - Disabling Watchdog...
Preloader - HW code:                    0x1208
Preloader - Target config:              0xe0
Preloader -     SBC enabled:            False
Preloader -     SLA enabled:            False
Preloader -     DAA enabled:            False
Preloader -     SWJTAG enabled:         False
Preloader -     EPP_PARAM at 0x600 after EMMC_BOOT/SDMMC_BOOT:  False
Preloader -     Root cert required:     False
Preloader -     Mem read auth:          True
Preloader -     Mem write auth:         True
Preloader -     Cmd 0xC8 blocked:       True
Preloader - Get Target info
Preloader - BROM mode detected.
Preloader -     HW subcode:             0x8a00
Preloader -     HW Ver:                 0xca00
Preloader -     SW Ver:                 0x0
Preloader - ME_ID:                      556B56CC473F6191882543187FDBE1FF
Preloader - SOC_ID:                     2DAE8FBBDB78BFB8BDA730338E236311B0BE5EF980A193D5BAD51CAD335B1EDA
DaHandler - Device is unprotected.
DaHandler - Device is in BROM-Mode. Bypassing security.
DaHandler
DaHandler - [LIB]: Failed to dump preloader from ram, provide a valid one via --preloader option
DAXML - Uploading xflash stage 1 from MTK_DA_V6.bin
XmlFlashExt - Patching da2 ...
XmlFlashExt - Patched read_register / write_register
XmlFlashExt - Patched write partitions / allow_read / allow_write
XmlFlashExt - Patched hash binding
DAXML - Successfully uploaded stage 1, jumping ..
Preloader - Jumping to 0x200000
Preloader - Jumping to 0x200000: ok.
DAXML - Stage 1 successfully loaded.
DAXML - Uploading stage 2...
Progress: |██████████| 100.0% Written (0x286/0x286, ) 0.00 MB/s
DAXML - Successfully uploaded stage 2.
DAXML - Successfully uploaded stage 2
DAXML - SLA is disabled
DAXML - DA XML Extensions successfully loaded.
DaHandler - Dumping partition "vbmeta_a"
Progress: |██████████| 100.0% Read (0x4000/0x4000, ) 6.19 MB/s62 MB/s
DaHandler - Patching vbmeta"
DaHandler - Patching verification + verity
DaHandler - Writing partition "vbmeta_a"
Progress: |██████████| 100.0% Written (0x4000/0x4000, ) 37.28 MB/s
DaHandler - Successfully patched vbmeta :)
```
